### ISSUE TITLE
TODO list for the clinical data management

### ISSUE THREAD

**Issue by wsy19961129** [ideafast/ideafast-portal#467]
### Things to determine ASAP (will update with the progress of the development)
 
- The features required for the backend data parser (e.g., data format, naming method of the field, etc.)
- Name format of the files of clinical data

### Features/Functions
- [ ] Data standards 
    - [ ]  Update the data parsers to fit the new data standards
- [ ] Database storage
    - [ ] A mechanism to store and trace multiple versions of data records
    - [ ] Increment loading support 
    - [ ] Need to determine the necessity to store query results in database (space vs time)
### APIs (partly already done)
- [ ] uploadFile: consider uploading clinical data whose file name is different from sensor data
- [ ] createDataCurationJob: allow to support multiple files; multiple behaviors based on the input record 
- [ ] createFieldCurationJob: allow to support multiple files
- [x] createProject: done
- [ ] getProject: need to check the $admin parameter at backend
- [ ] createQuery: create a query, but will not be executed immediately
- [ ] createQueryCurationJob: submit a job to execute a query to the job server. Note that the query result will not be returned
- [ ] getQueries: get all queries within a study/project
- [ ] getQueryById: get a single query, used for get the result of a single query
- [ ] updatePatientIdMapping: the patient mapping is generated the moment creating a project, but need to allow to update in case new data coming in (although it is not expected)

- [ ] Check all permissions
### UI

- [ ] Adjust old UI to fit IDEAFAST
- [ ] Datasets tab
    - [ ] Data management: update dataset summary/statistics caused by API changes
    - [ ] File Repository: update the table to fit the API change
    - [ ] Projects: summary/statistics for all projects (e.g., participants, project description, data updated date, queries, etc.) 
- [ ] Projects tab
    - [ ] Dashboard: complete the UI
    - [ ] Data: data structure; data fetcher/filter; results (statistics, sample data, visualization); functions to export query results as files
    - [ ] File Repository: update the table to fit the API change

**Comment by wsy19961129** [comment]
Regarding the data version and incremental loading, here some ideas:

As MongoDB has an upsert option (i.e., update if document found, otherwise insert it), I think it's better to use it directly.

**User behaviour**
1. We maintain multiple data versions at the same time;
2. The user needs to update the raw data (files) to the DMP; each file is mapped to a domain/path (e.g., MFI, body data);
3. The admin will select a time to parse all the data waiting to be parsed into the database (i.e., convert raw data to structured data);
4. The user can upload files at any time, but only after the admin allows to parse the data will such data be available for other users to read and obtain.

**How it works in MongoDB**
1. Each domain/path data of each participant is saved as an atomic data record (i.e., modification on any field of the data record will result in a new data record to be written in). Besides the data fields, each data record also has the following fields:

    - eid: denoting the subject
    - domain: denoting the domain/path of this record
    - study: denoting which study the record belongs to
    - version: denoting which version the record belongs to
    - job: denoting in which job the record is generated

2. When a new data record is parsed, there are two situations:
- The new data record matched an existing record (by comparing the eid and domain and all data fields), ignore it;
- Otherwise, insert it. Thus two kinds of records will be inserted: 1) a totally new record with a new eid-domain pair; 2) the eid-domain exists in the database but the values are modified.

3. How to obtain the data:
    We can group all records by the eid-domain pair, and find the latest records satisfied, i.e., the record whose versionid is the closest to the current data version of the study.

Here is an example of how the database will look like:

|index | Eid        | Domain           | Version  | Data... | Others... |
| ------------- |:-------------:|:-------------:|:-------------:|:-------------:| -----:|
| 0 | S1     | MFI | 1.0 |
| 1 | S1     | BD | 1.0 |
| 2 | S2      | BD      |   1.0 |
| 3 | S3 | MFI      |    1.0 |
| 4 | S1 | MFI      |    2.0 |
| 5 | S3 | MFI      |    2.0 |

If the current data version is 2.0, then the record of index 1,2 (from the closes version) 4,5 (from the current version) will be obtained, 2, 3 are substituted by 4, 5.
But if the current data version is 1.0, then 1,2,3,4 will be obtained.

**Disadvantages**
Once a data record is created, it cannot be deleted.
